core/src/main/java/io/questdb/griffin/SqlCodeGenerator.java

```diff
@@ -2079,53 +2080,72 @@ private RecordCursorFactory generateTableQuery(
                         }
-                        return new DataFrameRecordCursorFactory(myMeta, dfcFactory, rcf, orderByKeyColumn, filter, false, columnIndexes, columnSizes);
-                    }
 
-                    symbolValueList.clear();
+                        symbolValueList.clear();
 
-                    for (int i = 0, n = intrinsicModel.keyValues.size(); i < n; i++) {
-                        symbolValueList.add(intrinsicModel.keyValues.get(i));
-                    }
+                        for (int i = 0, n = intrinsicModel.keyValues.size(); i < n; i++) {
+                            symbolValueList.add(intrinsicModel.keyValues.get(i));
+                        }
 
-                    if (orderByKeyColumn) {
-                        myMeta.setTimestampIndex(-1);
-                        if (model.getOrderByDirectionAdvice().getQuick(0) == QueryModel.ORDER_DIRECTION_ASCENDING) {
-                            symbolValueList.sort(Chars.CHAR_SEQUENCE_COMPARATOR);
-                        } else {
-                            symbolValueList.sort(Chars.CHAR_SEQUENCE_COMPARATOR_DESC);
+                        if (orderByKeyColumn) {
+                            myMeta.setTimestampIndex(-1);
+                            if (model.getOrderByDirectionAdvice().getQuick(0) == QueryModel.ORDER_DIRECTION_ASCENDING) {
+                                symbolValueList.sort(Chars.CHAR_SEQUENCE_COMPARATOR);
+                            } else {
+                                symbolValueList.sort(Chars.CHAR_SEQUENCE_COMPARATOR_DESC);
+                            }
                         }
-                    }
+                        return new FilterOnValuesRecordCursorFactory(
+                                myMeta,
+                                dfcFactory,
+                                symbolValueList,
+                                keyColumnIndex,
+                                reader,
+                                filter,
+                                model.getOrderByAdviceMnemonic(),
+                                orderByKeyColumn,
+                                indexDirection,
+                                columnIndexes
+                        );
 
-                    return new FilterOnValuesRecordCursorFactory(
-                            myMeta,
-                            dfcFactory,
-                            symbolValueList,
-                            keyColumnIndex,
-                            reader,
-                            filter,
-                            model.getOrderByAdviceMnemonic(),
-                            orderByKeyColumn,
-                            indexDirection,
-                            columnIndexes
-                    );
+                    } else if (intrinsicModel.keyExcludedValues.size() > 0) {
+                        symbolValueList.clear();
+                        for (int i = 0, n = intrinsicModel.keyExcludedValues.size(); i < n; i++) {
```

Comment:
symboValueList is an ObjList and keyExcludedValues is CharSequenceHashSet. I could expose the underlying ObjList in CharSequenceHashSet, but that would break encapsulation

Reply:
nah, sorry, i thought it is the same type, ignore this please